# Exercise 4-8

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
# import dataframe
ords_prods_merge = pd.read_pickle(r"C:\Users\jinu5\Desktop\careerfoundry\Achivement 4\Instacart Basket Analysis_2024-01-02\02 Data\Prepared Data\ords_prods_merge.pkl")

In [3]:
df=ords_prods_merge[:1000000]

In [4]:
df.shape

(1000000, 18)

In [5]:
ords_prods_merge.head(10)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   
5   3367565        1             6                  2                  7   
6    550135        1             7                  1                  9   
7   3108588        1             8                  1                 14   
8   2295261        1             9                  1                 16   
9   2550362        1            10                  4                  8   

   days_since_previous_order  product_id  add_to_cart_order  reordered  \
0                        0.0         196                  1          0   
1                       15.0         196                  1          1   
2                       21.0         196                  1          1   
3                       29.0         196                  1          1   
4                       28.0         196                  1          1   
5                       19.0         196                  1          1   
6                       20.0         196                  1          1   
7                       14.0         196                  2          1   
8                        0.0         196                  4          1   
9                       30.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   
5         Soda        77              7     9.0   both  Mid-range product   
6         Soda        77              7     9.0   both  Mid-range product   
7         Soda        77              7     9.0   both  Mid-range product   
8         Soda        77              7     9.0   both  Mid-range product   
9         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy        Average orders  
1  Regularly busy    Slowest days        Average orders  
2  Regularly busy    Slowest days           Most orders  
3      Least busy    Slowest days        Average orders  
4      Least busy    Slowest days           Most orders  
5  Regularly busy  Regularly busy        Average orders  
6  Regularly busy    Busiest days        Average orders  
7  Regularly busy    Busiest days           Most orders  
8  Regularly busy    Busiest days           Most orders  
9      Least busy    Slowest days        Average orders

In [6]:
df.groupby('product_name')

In [7]:
df.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [8]:
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [9]:
df.groupby('department_id').order_number.mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [13]:
df.groupby('department_id').agg({'order_number':['mean', 'min','max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

## loyal customer - over 40: Loyal / 40<=&10>: Regular / 10< New

In [8]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\jinu5\AppData\Local\Temp\ipykernel_10600\1681540670.py:1: FutureWarning: The provided callable <function amax at 0x000002D17F965DA0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [15]:
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_previous_order  product_id  add_to_cart_order  reordered  \
0                        0.0         196                  1          0   
1                       15.0         196                  1          1   
2                       21.0         196                  1          1   
3                       29.0         196                  1          1   
4                       28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7       9   both  Mid-range product   
1         Soda        77              7       9   both  Mid-range product   
2         Soda        77              7       9   both  Mid-range product   
3         Soda        77              7       9   both  Mid-range product   
4         Soda        77              7       9   both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  max_order  
0  Regularly busy  Regularly busy        Average orders         10  
1  Regularly busy    Slowest days        Average orders         10  
2  Regularly busy    Slowest days           Most orders         10  
3      Least busy    Slowest days        Average orders         10  
4      Least busy    Slowest days           Most orders         10

In [16]:
ords_prods_merge.head(100)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2398795        1             2                  3                  7   
2     473747        1             3                  3                 12   
3    2254736        1             4                  4                  7   
4     431534        1             5                  4                 15   
..       ...      ...           ...                ...                ...   
95   3226575      360             1                  5                 12   
96   1469869      377             3                  5                 17   
97   1927023      387             2                  4                 10   
98    858092      420             4                  1                 19   
99    119314      420             5                  3                 19   

    days_since_previous_order  product_id  add_to_cart_order  reordered  \
0                         0.0         196                  1          0   
1                        15.0         196                  1          1   
2                        21.0         196                  1          1   
3                        29.0         196                  1          1   
4                        28.0         196                  1          1   
..                        ...         ...                ...        ...   
95                        0.0         196                  1          0   
96                        3.0         196                  9          0   
97                       22.0         196                  3          0   
98                       30.0         196                  2          0   
99                       16.0         196                  2          1   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7       9   both  Mid-range product   
1          Soda        77              7       9   both  Mid-range product   
2          Soda        77              7       9   both  Mid-range product   
3          Soda        77              7       9   both  Mid-range product   
4          Soda        77              7       9   both  Mid-range product   
..          ...       ...            ...     ...    ...                ...   
95         Soda        77              7       9   both  Mid-range product   
96         Soda        77              7       9   both  Mid-range product   
97         Soda        77              7       9   both  Mid-range product   
98         Soda        77              7       9   both  Mid-range product   
99         Soda        77              7       9   both  Mid-range product   

       busiest_day    busiest_days busiest_period_of_day  max_order  
0   Regularly busy  Regularly busy        Average orders         10  
1   Regularly busy    Slowest days        Average orders         10  
2   Regularly busy    Slowest days           Most orders         10  
3       Least busy    Slowest days        Average orders         10  
4       Least busy    Slowest days           Most orders         10  
..             ...             ...                   ...        ...  
95  Regularly busy  Regularly busy           Most orders          3  
96  Regularly busy  Regularly busy        Average orders          3  
97      Least busy    Slowest days           Most orders          8  
98  Regularly busy    Busiest days        Average orders         22  
99  Regularly busy    Slowest days        Average orders         22  

[100 rows x 19 columns]

In [17]:
pd.options.display.max_rows = None


In [18]:
ords_prods_merge.head(100)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2398795        1             2                  3                  7   
2     473747        1             3                  3                 12   
3    2254736        1             4                  4                  7   
4     431534        1             5                  4                 15   
5    3367565        1             6                  2                  7   
6     550135        1             7                  1                  9   
7    3108588        1             8                  1                 14   
8    2295261        1             9                  1                 16   
9    2550362        1            10                  4                  8   
10   2968173       15            15                  1                  9   
11   1870022       15            17                  2                 16   
12   1911383       15            18                  2                 11   
13   2715276       15            21                  1                  9   
14    487368       15            22                  1                 10   
15   2293453       19             2                  5                 14   
16   1973799       19             5                  6                 12   
17    532817       19             7                  4                 17   
18   1573906       21            10                  3                 10   
19   1593000       31            10                  3                  8   
20   2231262       31            17                  3                 11   
21   2580647       43             6                  4                 16   
22   2187180       43             9                  4                 12   
23   2497897       52             1                  1                  9   
24   1318871       52             2                  1                 10   
25   1261384       52             3                  1                 10   
26    580568       52             5                  2                 10   
27   2428073       52             7                  2                  7   
28    342306       52             9                  1                 11   
29   1520353       52            10                  2                 10   
30   2542086       52            11                  2                  9   
31    944694       52            12                  1                 16   
32   1268191       52            15                  2                 11   
33   1122089       52            17                  2                 10   
34   1498922       52            18                  2                 10   
35    180919       52            19                  3                  8   
36   2695875       52            21                  5                 10   
37   3205960       67             1                  3                 11   
38    970922       67             2                  1                 15   
39    596585       67             3                  1                 13   
40   2037087       67             4                  2                  9   
41   2474111       67             5                  3                 14   
42   1568481       67             6                  3                 12   
43   2048527       67             8                  2                 11   
44   2389031       67             9                  2                 14   
45   2378562       67            10                  1                 10   
46   1515698       67            11                  1                 11   
47   1928050       67            12                  2                 12   
48   1949274       67            13                  1                 10   
49    551869       67            14                  1                  9   
50    768572       67            16                  4                 10   


In [22]:
print.(ords_prods_merge(['user_id']=='81'))

SyntaxError: invalid syntax (634418661.py, line 1)

In [9]:
ords_prods_merge.loc[ords_prods_merge['max_order']>40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order']>10), 'loyalty_flag' ] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order']<= 10, 'loyalty_flag']='New customer'
                                       

In [10]:
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [29]:
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

In [31]:
ords_prods_merge.to_pickle(r"C:\Users\jinu5\Desktop\careerfoundry\Achivement 4\Instacart Basket Analysis_2024-01-02\02 Data\Prepared Data\ords_prods_merge.pkl")

# Task 4-8

### Step 2.  find the aggregated mean of the “order_number” column grouped by “department_id”

In [11]:
ords_prods_merge.groupby('department_id').agg({'order_number':['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.457838   1  99
2                17.277920   1  99
3                17.170395   1  99
4                17.811403   1  99
5                15.215751   1  99
6                16.439806   1  99
7                17.225802   1  99
8                15.340650   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.170638   1  99
12               15.887671   1  99
13               16.583536   1  99
14               16.773669   1  99
15               16.165037   1  99
16               17.665606   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

### Step 3. Analyze the result

Subse /  whole dataframe

 - count of department: 7 / 21
 - higest mean of department : 16,  19.463012  / 21, 22.902379
 - lowest mean of department: 17,  11.22.902379 / 5, 15.215751

department_id  
4     18.825780  
7     17.472355  
13    17.993423  
14    19.246334  
16    19.463012  
17    11.294069  
19    19.305237  
20    17.599636  

### Step 4. creating a loyalty flag for existing customers using the transform() and loc() functions.

In [12]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\jinu5\AppData\Local\Temp\ipykernel_10600\1681540670.py:1: FutureWarning: The provided callable <function amax at 0x000002D17F965DA0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [13]:
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_previous_order  product_id  add_to_cart_order  reordered  \
0                        0.0         196                  1          0   
1                       15.0         196                  1          1   
2                       21.0         196                  1          1   
3                       29.0         196                  1          1   
4                       28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy        Average orders         10   
1  Regularly busy    Slowest days        Average orders         10   
2  Regularly busy    Slowest days           Most orders         10   
3      Least busy    Slowest days        Average orders         10   
4      Least busy    Slowest days           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

In [14]:
ords_prods_merge.loc[ords_prods_merge['max_order']>40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order']>10), 'loyalty_flag' ] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order']<= 10, 'loyalty_flag']='New customer'

In [15]:
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

### Step 5. difference between the spending habits of the three types of customers

In [16]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices':['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

Loyal customer's average spend is lower than Regular and New customer types. The hiesst spend by customer type is New customer. 

### Step 6. different types of spenders. 10>= : high spender / 10< : Low spender. 

In [17]:
# each user's average price across all their orders, spending flag

ords_prods_merge['average_spending']=ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\jinu5\AppData\Local\Temp\ipykernel_10600\2097908879.py:3: FutureWarning: The provided callable <function mean at 0x000002D17F966AC0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['average_spending']=ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [15]:
ords_prods_merge.tail()

order_id  user_id  order_number  order_day_of_week  \
32404854   1320836   202557            17                  2   
32404855     31526   202557            18                  5   
32404856    758936   203436             1                  2   
32404857   2745165   203436             2                  3   
32404858   3093936   205420             1                  4   

          order_hour_of_day  days_since_previous_order  product_id  \
32404854                 15                        1.0       43553   
32404855                 11                        3.0       43553   
32404856                  7                        0.0       42338   
32404857                  5                       15.0       42338   
32404858                 14                        0.0       28818   

          add_to_cart_order  reordered                   product_name  ...  \
32404854                  2          1            Orange Energy Shots  ...   
32404855                  2          1            Orange Energy Shots  ...   
32404856                  4          0          Zucchini Chips, Pesto  ...   
32404857                 16          1          Zucchini Chips, Pesto  ...   
32404858                  8          0  Hot Oatmeal Multigrain Raisin  ...   

          department_id  prices  _merge    price_range_loc     busiest_day  \
32404854              7       3    both  Low_range_product  Regularly busy   
32404855              7       3    both  Low_range_product  Regularly busy   
32404856             19       6    both  Mid-range product  Regularly busy   
32404857             19       6    both  Mid-range product  Regularly busy   
32404858             14      10    both  Mid-range product      Least busy   

            busiest_days busiest_period_of_day max_order      loyalty_flag  \
32404854  Regularly busy           Most orders        31  Regular customer   
32404855  Regularly busy           Most orders        31  Regular customer   
32404856  Regularly busy        Average orders         3      New customer   
32404857    Slowest days         Fewest orders         3      New customer   
32404858    Slowest days           Most orders        16  Regular customer   

         average_spending  
32404854         6.476190  
32404855         6.476190  
32404856         7.131579  
32404857         7.131579  
32404858         7.228814  

[5 rows x 21 columns]

In [16]:
ords_prods_merge['sum_spending']=ords_prods_merge.groupby(['user_id'])['prices'].transform(np.sum)

C:\Users\jinu5\AppData\Local\Temp\ipykernel_25304\1598272054.py:1: FutureWarning: The provided callable <function sum at 0x0000024ED8F02700> is currently using SeriesGroupBy.sum. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "sum" instead.
  ords_prods_merge['sum_spending']=ords_prods_merge.groupby(['user_id'])['prices'].transform(np.sum)


In [17]:
ords_prods_merge.tail()

order_id  user_id  order_number  order_day_of_week  \
32404854   1320836   202557            17                  2   
32404855     31526   202557            18                  5   
32404856    758936   203436             1                  2   
32404857   2745165   203436             2                  3   
32404858   3093936   205420             1                  4   

          order_hour_of_day  days_since_previous_order  product_id  \
32404854                 15                        1.0       43553   
32404855                 11                        3.0       43553   
32404856                  7                        0.0       42338   
32404857                  5                       15.0       42338   
32404858                 14                        0.0       28818   

          add_to_cart_order  reordered                   product_name  ...  \
32404854                  2          1            Orange Energy Shots  ...   
32404855                  2          1            Orange Energy Shots  ...   
32404856                  4          0          Zucchini Chips, Pesto  ...   
32404857                 16          1          Zucchini Chips, Pesto  ...   
32404858                  8          0  Hot Oatmeal Multigrain Raisin  ...   

          prices  _merge    price_range_loc     busiest_day    busiest_days  \
32404854       3    both  Low_range_product  Regularly busy  Regularly busy   
32404855       3    both  Low_range_product  Regularly busy  Regularly busy   
32404856       6    both  Mid-range product  Regularly busy  Regularly busy   
32404857       6    both  Mid-range product  Regularly busy    Slowest days   
32404858      10    both  Mid-range product      Least busy    Slowest days   

         busiest_period_of_day max_order      loyalty_flag  average_spending  \
32404854           Most orders        31  Regular customer          6.476190   
32404855           Most orders        31  Regular customer          6.476190   
32404856        Average orders         3      New customer          7.131579   
32404857         Fewest orders         3      New customer          7.131579   
32404858           Most orders        16  Regular customer          7.228814   

         sum_spending  
32404854         2176  
32404855         2176  
32404856          271  
32404857          271  
32404858          853  

[5 rows x 22 columns]

In [18]:
ords_prods_merge.loc[ords_prods_merge['average_spending']>=10, 'spending_flag'] = 'High spender'

In [19]:
ords_prods_merge.loc[ords_prods_merge['average_spending'] < 10, 'spending_flag'] = 'Low spender'

In [20]:
ords_prods_merge.tail()

order_id  user_id  order_number  order_day_of_week  \
32404854   1320836   202557            17                  2   
32404855     31526   202557            18                  5   
32404856    758936   203436             1                  2   
32404857   2745165   203436             2                  3   
32404858   3093936   205420             1                  4   

          order_hour_of_day  days_since_previous_order  product_id  \
32404854                 15                        1.0       43553   
32404855                 11                        3.0       43553   
32404856                  7                        0.0       42338   
32404857                  5                       15.0       42338   
32404858                 14                        0.0       28818   

          add_to_cart_order  reordered                   product_name  ...  \
32404854                  2          1            Orange Energy Shots  ...   
32404855                  2          1            Orange Energy Shots  ...   
32404856                  4          0          Zucchini Chips, Pesto  ...   
32404857                 16          1          Zucchini Chips, Pesto  ...   
32404858                  8          0  Hot Oatmeal Multigrain Raisin  ...   

          prices  _merge    price_range_loc     busiest_day    busiest_days  \
32404854     3.7    both  Low_range_product  Regularly busy  Regularly busy   
32404855     3.7    both  Low_range_product  Regularly busy  Regularly busy   
32404856     6.9    both  Mid-range product  Regularly busy  Regularly busy   
32404857     6.9    both  Mid-range product  Regularly busy    Slowest days   
32404858    10.3    both  Mid-range product      Least busy    Slowest days   

         busiest_period_of_day max_order      loyalty_flag  average_spending  \
32404854           Most orders        31  Regular customer          6.905655   
32404855           Most orders        31  Regular customer          6.905655   
32404856        Average orders         3      New customer          7.631579   
32404857         Fewest orders         3      New customer          7.631579   
32404858           Most orders        16  Regular customer          7.684746   

         spending_flag  
32404854   Low spender  
32404855   Low spender  
32404856   Low spender  
32404857   Low spender  
32404858   Low spender  

[5 rows x 22 columns]

In [21]:
ords_prods_merge['spending_flag'].value_counts(dropna=False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

### Step 7. frequent versus non-frequent customers, creating order frequency flag, according to the median in the “days_since_prior_order”

In [22]:
ords_prods_merge['order_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_previous_order'].transform(np.median)

C:\Users\jinu5\AppData\Local\Temp\ipykernel_10600\2655326495.py:1: FutureWarning: The provided callable <function median at 0x000002D17FAA3A60> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['order_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_previous_order'].transform(np.median)


In [22]:
ords_prods_merge.tail()

order_id  user_id  order_number  order_day_of_week  \
32404854   1320836   202557            17                  2   
32404855     31526   202557            18                  5   
32404856    758936   203436             1                  2   
32404857   2745165   203436             2                  3   
32404858   3093936   205420             1                  4   

          order_hour_of_day  days_since_previous_order  product_id  \
32404854                 15                        1.0       43553   
32404855                 11                        3.0       43553   
32404856                  7                        0.0       42338   
32404857                  5                       15.0       42338   
32404858                 14                        0.0       28818   

          add_to_cart_order  reordered                   product_name  ...  \
32404854                  2          1            Orange Energy Shots  ...   
32404855                  2          1            Orange Energy Shots  ...   
32404856                  4          0          Zucchini Chips, Pesto  ...   
32404857                 16          1          Zucchini Chips, Pesto  ...   
32404858                  8          0  Hot Oatmeal Multigrain Raisin  ...   

            price_range_loc     busiest_day    busiest_days  \
32404854  Low_range_product  Regularly busy  Regularly busy   
32404855  Low_range_product  Regularly busy  Regularly busy   
32404856  Mid-range product  Regularly busy  Regularly busy   
32404857  Mid-range product  Regularly busy    Slowest days   
32404858  Mid-range product      Least busy    Slowest days   

         busiest_period_of_day max_order      loyalty_flag average_spending  \
32404854           Most orders        31  Regular customer         6.476190   
32404855           Most orders        31  Regular customer         6.476190   
32404856        Average orders         3      New customer         7.131579   
32404857         Fewest orders         3      New customer         7.131579   
32404858           Most orders        16  Regular customer         7.228814   

         sum_spending  spending_flag order_frequency  
32404854         2176    Low spender             7.0  
32404855         2176    Low spender             7.0  
32404856          271    Low spender            15.0  
32404857          271    Low spender            15.0  
32404858          853    Low spender            12.0  

[5 rows x 24 columns]

In [23]:
ords_prods_merge.loc[ords_prods_merge['order_frequency']>20, 'frequency_flag'] = 'Non-frequency customer'

In [24]:
ords_prods_merge.loc[ords_prods_merge['order_frequency']<=10, 'frequency_flag'] = 'Frequent customer'

In [25]:
ords_prods_merge.loc[(ords_prods_merge['order_frequency']<=20) & (ords_prods_merge['order_frequency'] > 10) , 'frequency_flag'] = 'RegularF customer'

In [26]:
ords_prods_merge.tail()

order_id  user_id  order_number  order_day_of_week  \
32404854   1320836   202557            17                  2   
32404855     31526   202557            18                  5   
32404856    758936   203436             1                  2   
32404857   2745165   203436             2                  3   
32404858   3093936   205420             1                  4   

          order_hour_of_day  days_since_previous_order  product_id  \
32404854                 15                        1.0       43553   
32404855                 11                        3.0       43553   
32404856                  7                        0.0       42338   
32404857                  5                       15.0       42338   
32404858                 14                        0.0       28818   

          add_to_cart_order  reordered                   product_name  ...  \
32404854                  2          1            Orange Energy Shots  ...   
32404855                  2          1            Orange Energy Shots  ...   
32404856                  4          0          Zucchini Chips, Pesto  ...   
32404857                 16          1          Zucchini Chips, Pesto  ...   
32404858                  8          0  Hot Oatmeal Multigrain Raisin  ...   

            price_range_loc     busiest_day    busiest_days  \
32404854  Low_range_product  Regularly busy  Regularly busy   
32404855  Low_range_product  Regularly busy  Regularly busy   
32404856  Mid-range product  Regularly busy  Regularly busy   
32404857  Mid-range product  Regularly busy    Slowest days   
32404858  Mid-range product      Least busy    Slowest days   

         busiest_period_of_day max_order      loyalty_flag average_spending  \
32404854           Most orders        31  Regular customer         6.905655   
32404855           Most orders        31  Regular customer         6.905655   
32404856        Average orders         3      New customer         7.631579   
32404857         Fewest orders         3      New customer         7.631579   
32404858           Most orders        16  Regular customer         7.684746   

         spending_flag  order_frequency     frequency_flag  
32404854   Low spender              7.0  Frequent customer  
32404855   Low spender              7.0  Frequent customer  
32404856   Low spender             15.0  RegularF customer  
32404857   Low spender             15.0  RegularF customer  
32404858   Low spender             12.0  RegularF customer  

[5 rows x 24 columns]

In [27]:
ords_prods_merge['frequency_flag'].value_counts(dropna=False)

frequency_flag
Frequent customer         22796659
RegularF customer          6921472
Non-frequency customer     2686728
Name: count, dtype: int64

### Export dataframe

In [ ]:
ords_prods_merge.to_pickle(r"C:\Users\jinu5\Desktop\careerfoundry\Achivement 4\Instacart Basket Analysis_2024-01-02\02 Data\Prepared Data\ords_prods_merge.pkl")